File: ts/src/audit.test.ts
import { describe, it } from "node:test";
import assert from "node:assert/strict";
import { mkdtempSync, readFileSync, rmSync } from "node:fs";
import { tmpdir } from "node:os";
import { join } from "node:path";
import { AuditLogger } from "./audit.js";

describe("AuditLogger", () => {
  it("writes a well-formed ALLOW entry", () => {
    const dir = mkdtempSync(join(tmpdir(), "audit-"));
    const path = join(dir, "audit.log");
    const logger = new AuditLogger(path);
    logger.allow("GET", "/_ping");

    const lines = readFileSync(path, "utf8").trim().split("\n");
    assert.equal(lines.length, 1);
    const entry = JSON.parse(lines[0]);
    assert.equal(entry.decision, "ALLOW");
    assert.equal(entry.method, "GET");
    assert.equal(entry.uri, "/_ping");
    assert.equal(entry.reason, "");
    assert.ok(entry.request_id);
    assert.ok(entry.timestamp);
    rmSync(dir, { recursive: true, force: true });
  });

  it("writes a well-formed DENY entry with reason", () => {
    const dir = mkdtempSync(join(tmpdir(), "audit-"));
    const path = join(dir, "audit.log");
    const logger = new AuditLogger(path);
    logger.deny("POST", "/containers/create", "exec denied by policy");

    const lines = readFileSync(path, "utf8").trim().split("\n");
    assert.equal(lines.length, 1);
    const entry = JSON.parse(lines[0]);
    assert.equal(entry.decision, "DENY");
    assert.equal(entry.method, "POST");
    assert.equal(entry.uri, "/containers/create");
    assert.equal(entry.reason, "exec denied by policy");
    rmSync(dir, { recursive: true, force: true });
  });

  it("appends one JSON line per event", () => {
    const dir = mkdtempSync(join(tmpdir(), "audit-"));
    const path = join(dir, "audit.log");
    const logger = new AuditLogger(path);
    for (let i = 0; i < 5; i++) {
      logger.allow("GET", "/version");
    }

    const lines = readFileSync(path, "utf8").trim().split("\n");
    assert.equal(lines.length, 5);
    for (const line of lines) {
      const entry = JSON.parse(line);
      assert.equal(entry.decision, "ALLOW");
    }
    rmSync(dir, { recursive: true, force: true });
  });

  it("never throws when the log file cannot be written", () => {
    const logger = new AuditLogger("/nonexistent-dir/audit.log");
    assert.doesNotThrow(() => logger.allow("GET", "/_ping"));
  });
});

File: go/internal/audit/audit_test.go
package audit

import (
	"bufio"
	"encoding/json"
	"os"
	"path/filepath"
	"testing"
)

func readEntries(t *testing.T, path string) []map[string]interface{} {
	t.Helper()
	f, err := os.Open(path)
	if err != nil {
		t.Fatalf("failed to open audit log: %v", err)
	}
	defer f.Close()

	var entries []map[string]interface{}
	scanner := bufio.NewScanner(f)
	for scanner.Scan() {
		var entry map[string]interface{}
		if err := json.Unmarshal(scanner.Bytes(), &entry); err != nil {
			t.Fatalf("invalid audit JSON line: %v", err)
		}
		entries = append(entries, entry)
	}
	if err := scanner.Err(); err != nil {
		t.Fatalf("failed to read audit log: %v", err)
	}
	return entries
}

func TestLogger_AllowWritesEntry(t *testing.T) {
	path := filepath.Join(t.TempDir(), "audit.log")
	l, err := NewLogger(path)
	if err != nil {
		t.Fatalf("NewLogger failed: %v", err)
	}
	extra := map[string]interface{}{"service": "beacon", "image": "chainsafe/lodestar:next"}
	l.Allow("GET", "/_ping", "request allowed", extra)
	l.Close()

	entries := readEntries(t, path)
	if len(entries) != 1 {
		t.Fatalf("expected 1 entry, got %d", len(entries))
	}
	e := entries[0]
	assertField(t, e, "decision", "ALLOW")
	assertField(t, e, "method", "GET")
	assertField(t, e, "uri", "/_ping")
	assertField(t, e, "reason", "request allowed")
	if extra, ok := e["extra"].(map[string]interface{}); !ok {
		t.Fatalf("expected extra object, got %T", e["extra"])
	} else {
		if extra["service"] != "beacon" {
			t.Fatalf("expected extra.service beacon, got %v", extra["service"])
		}
		if extra["image"] != "chainsafe/lodestar:next" {
			t.Fatalf("expected extra.image, got %v", extra["image"])
		}
	}
	if e["request_id"] == "" {
		t.Fatal("expected non-empty request_id")
	}
	if e["timestamp"] == "" {
		t.Fatal("expected non-empty timestamp")
	}
}

func TestLogger_DenyWritesEntry(t *testing.T) {
	path := filepath.Join(t.TempDir(), "audit.log")
	l, err := NewLogger(path)
	if err != nil {
		t.Fatalf("NewLogger failed: %v", err)
	}
	l.Deny("POST", "/containers/create", "exec denied by policy", nil)
	l.Close()

	entries := readEntries(t, path)
	if len(entries) != 1 {
		t.Fatalf("expected 1 entry, got %d", len(entries))
	}
	e := entries[0]
	assertField(t, e, "decision", "DENY")
	assertField(t, e, "method", "POST")
	assertField(t, e, "uri", "/containers/create")
	assertField(t, e, "reason", "exec denied by policy")
}

func TestLogger_MultipleEntries(t *testing.T) {
	path := filepath.Join(t.TempDir(), "audit.log")
	l, err := NewLogger(path)
	if err != nil {
		t.Fatalf("NewLogger failed: %v", err)
	}
	for i := 0; i < 5; i++ {
		l.Allow("GET", "/version", "request allowed", nil)
	}
	l.Close()

	entries := readEntries(t, path)
	if len(entries) != 5 {
		t.Fatalf("expected 5 entries, got %d", len(entries))
	}
}

func TestLogger_NopLoggerWritesNowhere(t *testing.T) {
	l := NewNopLogger()
	l.Allow("GET", "/_ping", "request allowed", nil)
	l.Deny("POST", "/containers/create", "denied", nil)
	l.Close()
}

func assertField(t *testing.T, e map[string]interface{}, key, expected string) {
	t.Helper()
	if got, ok := e[key].(string); !ok || got != expected {
		t.Fatalf("expected %s=%q, got %v", key, expected, e[key])
	}
}